include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+
+#ifdef _KERNEL_MODE
+#error This header is not supported in kernel-mode.
+#endif
+
+#include <new.h> // new(std::nothrow)
+#include "wistd_type_traits.h"
+#include <OleAuto.h>
+#include "resource.h"
+
+namespace wil
+{
+#if !defined(__WIL_SAFEARRAY_)
+#define __WIL_SAFEARRAY_
+    /// @cond
+    namespace details
+    {
+        template<typename T> struct infer_safearray_type_traits{};
+        template<> struct infer_safearray_type_traits<char>                       { static constexpr auto vartype = VT_I1; };
+        //template<> struct infer_safearray_type_traits<short>                      { static constexpr auto vartype = VT_I2; };

```

Comment:
What's the reason for this and `double` to be commented out? Does `I2` & `R8` not exist? Comments about why they are not present are more helpful than commented out code

Reply:
I2 and R8 are tricky to support because they're the same as other types as far as the compiler is concerned.  `short` is the same as `BSTR` and `double` is the same as `DATE`.

Up at the top Chris is suggesting using some dedicated enums for each type that could be confused so they can be explicit.  I'll look into that.

Reply:
Again, comments about _why_ these types are not present are more helpful than leaving in commented out code. I see the `DATE`/`double` conflict, however `BSTR` is defined as a pointer and shouldn't conflict with `short` - did you mean something else?

But now that I know why they are not present, what's the behavior if I try and use `double` with this code? Does it give me a compilation error? Does it succeed, but silently give me incorrect behavior because values are interpreted incorrectly? Or does it end up working correctly? This is all information that is more useful to have commented in the code than commented out lines like this.